test/e2e/restart.go

```diff
@@ -0,0 +1,367 @@
+	failed := false
+	// TODO(mbforbes): Change to `for range` syntax once we support only Go
+	// >= 1.4.
+	for i := range nodeList.Items {
+		_ = i
+		if !<-result {
+			failed = true
+		}
+	}
+	if failed {
+		return nodeNames, fmt.Errorf("at least one node failed to be ready")
+	}
+	return nodeNames, nil
+}
+
+// restartNodes uses provider to do a restart of all nodes in the cluster,
+// allowing up to nt per node.
+func restartNodes(provider string, nt time.Duration) error {
+	switch provider {
+	case "gce":
+		return migRollingUpdate(nt)
+	default:
+		return fmt.Errorf("restartNodes(...) not implemented for %s", provider)
+	}
+}
+
+// migRollingUpdate starts a MIG rolling update and waits up to nt times the
+// nubmer of nodes for it to complete.
+func migRollingUpdate(nt time.Duration) error {
+	By("getting the name of the template for the managed instance group")
+	templ, err := migTemplate()
+	if err != nil {
+		return fmt.Errorf("couldn't get MIG template name: %v", err)
+	}
+
+	By("starting the managed instance group rolling update")
+	id, err := migRollingUpdateStart(templ, nt)
+	if err != nil {
+		return fmt.Errorf("couldn't start the MIG rolling update: %v", err)
+	}
+
+	By("polling the managed instance group rolling update until it completes")
+	if err := migRollingUpdatePoll(id, nt); err != nil {
+		return fmt.Errorf("err waiting until update completed: %v", err)
+	}
+
+	return nil
+}
+
+// migTemlate (GCE/GKE-only) returns the name of the MIG template that the
+// nodes of the cluster use.
+func migTemplate() (string, error) {
+	var err error
+	var key = "instanceTemplate"
+	for i := 0; i < singleCallAttempts; i++ {
+		var o []byte
+		// TODO(mbforbes): make this hit the compute API directly instead of
+		// shelling out to gcloud.
+		o, err = exec.Command("gcloud", "preview", "managed-instance-groups",
+			fmt.Sprintf("--project=%s", testContext.CloudConfig.ProjectID),
+			fmt.Sprintf("--zone=%s", testContext.CloudConfig.Zone),
+			"describe",
+			testContext.CloudConfig.NodeInstanceGroup).CombinedOutput()
+		if err != nil {
+			continue
+		}
+		output := string(o)
+
+		// The 'describe' call probably succeeded; parse the output and try to
+		// find the line that looks like "instanceTemplate: url/to/<templ>" and
+		// return <templ>.
+		if val := parseKVLines(output, key); len(val) > 0 {
+			url := strings.Split(val, "/")
+			templ := url[len(url)-1]
+			Logf("MIG group %s using template: %s", testContext.CloudConfig.NodeInstanceGroup, templ)
+			return templ, nil
+		}
+		err = fmt.Errorf("couldn't find %s in output to get MIG template. Output: %s", key, output)
+	}
+	return "", err
+}
+
+// migRollingUpdateStart (GCE/GKE-only) starts a MIG rolling update using templ
+// as the new template, waiting up to nt per node, and returns the ID of that
+// update.
+func migRollingUpdateStart(templ string, nt time.Duration) (string, error) {
+	var err error
+	prefix, suffix := "Started [", "]."
+	for i := 0; i < singleCallAttempts; i++ {
```

Comment:
Question: I'm doing something different for retries on single API calls versus "timeout"/"waiting" calls, where a single call is tried n times. This seemed slightly better choosing timeouts that all single calls are supposed to conform to. Do we really want to switch all of these to waiting for time periods? If so, what's a good periods? Single API calls *should* fail very quickly, so we could get all 5 failures here in ~2-3 seconds, but might that be too risky? If we have it be 20 seconds or 1 minute, we might call a few hundred times; is that what we want?